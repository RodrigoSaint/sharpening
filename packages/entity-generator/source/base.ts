export type FieldType = "string" | "int" | "decimal" | "date";

export interface Field {
  name: string;
  isRequired: boolean;
  type: FieldType;
}

export type EntityAction =
  | "pagination"
  | "list"
  | "detail"
  | "create"
  | "update"
  | "delete";

export interface Entity {
  name: string;
  fieldCollection: Field[];
  actionCollection?: EntityAction[];
  formFieldCollection?: string[];
  displayFieldCollection?: string[];
}
