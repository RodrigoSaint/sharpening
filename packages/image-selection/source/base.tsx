import React, { useState } from "react";
import Modal from "@rodrigosaint/modal";
import ManagedImage from "@rodrigosaint/managed-image";

import ImageSelectionModalContent, {
  ImageSelectionModalContentProps,
} from "./content";

export interface ImageSelectionProps<T>
  extends Omit<ImageSelectionModalContentProps<T>, "close"> {
  renderWithoutImage: () => React.ReactNode;
  getImageUrl: (item: T) => string;
  header: string | React.ReactNode;
  initialValue?: T;
}

export default function ImageSelection<T>(props: ImageSelectionProps<T>) {
  const [src, setSrc] = useState<string>(
    props.initialValue ? props.getImageUrl(props.initialValue) : undefined
  );

  const onSelected = React.useCallback(
    (item: T) => {
      setSrc(props.getImageUrl(item));
      props.onSelected(item);
    },
    [props.onSelected, props.getImageUrl]
  );

  return (
    <>
      <Modal
        renderTrigger={() => (
          <ManagedImage
            src={src}
            renderWithoutImage={props.renderWithoutImage}
            renderOnError={props.renderWithoutImage}
          />
        )}
        renderContent={(close) => (
          <ImageSelectionModalContent
            {...props}
            onSelected={onSelected}
            close={close}
          />
        )}
        header={props.header}
      />
    </>
  );
}
