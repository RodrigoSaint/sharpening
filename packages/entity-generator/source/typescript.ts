import { Entity, Field, FieldType } from "./base";
import { getCamel, getCapital } from "./helper";

function getFieldType(type: FieldType) {
  const map: Record<FieldType, string> = {
    string: "string",
    int: "number",
    decimal: "number",
    date: "Date",
  };
  return map[type];
}

export function getField(field: Field) {
  return `\t${field.name}: ${getFieldType(field.type)};\n`;
}

export function getInterface(entity: Entity) {
  return `interface ${getCapital(entity.name)} {\n${entity.fieldCollection
    .map(getField)
    .join("")}}`;
}

export function getFieldValidation(field: Field) {
  const map: Record<FieldType, string> = {
    string: ".string()",
    int: ".number()",
    decimal: ".number()",
    date: ".date()",
  };

  const isRequired = field.isRequired ? ".required()" : "";

  return `\t${field.name}: yup${map[field.type]}${isRequired}`;
}

export function getEntityValidation(entity: Entity) {
  return `const ${getCamel(
    entity.name
  )}Schema = yup.object().shape({\n${entity.fieldCollection
    .map(getFieldValidation)
    .join(",\n")}\n})`;
}
