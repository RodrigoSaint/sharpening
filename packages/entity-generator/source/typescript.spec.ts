import { Entity } from "./base";
import { getInterface, getEntityValidation } from "./typescript";
const entity: Entity = {
  name: "User",
  actionCollection: [
    "pagination",
    "list",
    "detail",
    "create",
    "update",
    "delete",
  ],
  fieldCollection: [
    {
      name: "name",
      isRequired: true,
      type: "string",
    },
    {
      name: "email",
      isRequired: false,
      type: "string",
    },
    {
      name: "age",
      isRequired: false,
      type: "int",
    },
  ],
};

describe("Typescript", () => {
  it("Gets interface", () => {
    expect(getInterface(entity)).toMatchSnapshot("getInterface");
  });
  it("Get entity validation", () => {
    expect(getEntityValidation(entity)).toMatchSnapshot("getEntityValidation");
  });
});
