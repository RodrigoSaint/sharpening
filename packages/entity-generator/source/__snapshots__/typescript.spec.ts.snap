// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript Get entity validation: getEntityValidation 1`] = `
"const userSchema = yup.object().shape({
	name: yup.string().required(),
	email: yup.string(),
	age: yup.number()
})"
`;

exports[`Typescript Gets interface: getInterface 1`] = `
"interface User {
	name: string;
	email: string;
	age: number;
}"
`;
